File: src/test/test_arith.py
#!/usr/bin/env python3
# Arbitrary-precision oracle for test_arith.cpp output.
import sys

M256 = (1 << 256) - 1
M512 = (1 << 512) - 1
fails = 0
n = 0

def set_compact(nCompact):
    size = nCompact >> 24
    word = nCompact & 0x007fffff
    if size <= 3:
        word >>= 8 * (3 - size)
        val = word
    else:
        val = word << (8 * (size - 3))
    neg = (word != 0) and (nCompact & 0x00800000) != 0
    over = (word != 0) and (size > 34 or (word > 0xff and size > 33) or (word > 0xffff and size > 32))
    return val, neg, over

def get_compact(val, fNegative=False):
    size = (val.bit_length() + 7) // 8
    if size <= 3:
        compact = (val & ((1 << 64) - 1)) << (8 * (3 - size))
    else:
        compact = (val >> (8 * (size - 3))) & ((1 << 64) - 1)
    if compact & 0x00800000:
        compact >>= 8
        size += 1
    compact |= size << 24
    if fNegative and (compact & 0x007fffff):
        compact |= 0x00800000
    return compact & 0xffffffff

for line in sys.stdin:
    p = line.split()
    if not p:
        continue
    k = p[0]
    n += 1
    if k == "MUL":
        a, b, r = int(p[1], 16), int(p[2], 16), int(p[3], 16)
        want = (a * b) & M256
        if want != r:
            print("MUL mismatch", hex(a), hex(b), "want", hex(want), "got", hex(r)); fails += 1
    elif k == "MULW":
        a, b, r = int(p[1], 16), int(p[2], 16), int(p[3], 16)
        want = (a * b) & M512
        if want != r:
            print("MULW mismatch", hex(a), hex(b), "want", hex(want), "got", hex(r)); fails += 1
    elif k == "DIV":
        a, b, r = int(p[1], 16), int(p[2], 16), int(p[3], 16)
        want = a // b
        if want != r:
            print("DIV mismatch", hex(a), hex(b), "want", hex(want), "got", hex(r)); fails += 1
    elif k == "SETCOMPACT":
        nbits = int(p[1], 16)
        val, neg, over = set_compact(nbits)
        gotval = int(p[2], 16)
        gotneg, gotover = int(p[3]), int(p[4])
        if gotover != int(over):
            print("SETCOMPACT overflow flag", hex(nbits), "want", over, "got", gotover); fails += 1
        elif not over and (val & M256) != gotval:
            print("SETCOMPACT value", hex(nbits), "want", hex(val & M256), "got", hex(gotval)); fails += 1
        if gotneg != int(neg):
            print("SETCOMPACT neg flag", hex(nbits), "want", neg, "got", gotneg); fails += 1
    elif k == "GETCOMPACT":
        val = int(p[1], 16)
        got = int(p[2], 16)
        want = get_compact(val, False)
        if want != got:
            print("GETCOMPACT", hex(val), "want", hex(want), "got", hex(got)); fails += 1
    elif k == "RETARGET":
        tgt, m, d, r = int(p[1], 16), int(p[2]), int(p[3]), int(p[4], 16)
        want = ((tgt * m) // d) & M256
        if want != r:
            print("RETARGET mismatch tgt", hex(tgt), "m", m, "d", d, "want", hex(want), "got", hex(r)); fails += 1
    elif k == "TRUST":
        tgt, pos, pw = int(p[1], 16), int(p[2], 16), int(p[3], 16)
        want_pos = ((1 << 256) // (tgt + 1)) & M256
        want_pw = (((M256 >> 8) + 1 - 1) // (tgt + 1))  # powLimit = ~0 >> 8
        want_pw = ((M256 >> 8) // (tgt + 1))
        if want_pos != pos:
            print("TRUST pos", hex(tgt), "want", hex(want_pos), "got", hex(pos)); fails += 1
        if want_pw != pw:
            print("TRUST pow", hex(tgt), "want", hex(want_pw), "got", hex(pw)); fails += 1

print(f"checked {n} vectors, {fails} failures")
sys.exit(1 if fails else 0)

File: src/test/test_arith.cpp
// Throwaway: emit test vectors for the new fixed-width uint256/uint512
// arithmetic in uint256.h so an arbitrary-precision oracle (test_arith.py)
// can confirm it matches what CBigNum used to compute. Not part of any build.
//
//   g++ -std=c++11 -O2 -I. test_arith.cpp -o /tmp/ta && /tmp/ta | python3 test_arith.py

#include <cstdio>
#include <stdint.h>

// uint256.h wants this (normally in util.cpp)
const signed char p_util_hexdigit[256] =
{ -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
   0, 1, 2, 3, 4, 5, 6, 7, 8, 9,-1,-1,-1,-1,-1,-1,
  -1,0xa,0xb,0xc,0xd,0xe,0xf,-1,-1,-1,-1,-1,-1,-1,-1,-1,
  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
  -1,0xa,0xb,0xc,0xd,0xe,0xf,-1,-1,-1,-1,-1,-1,-1,-1,-1,
  -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1, };

#include "uint256.h"

static uint64_t s = 88172645463325252ULL;
static uint64_t xr() { s ^= s << 13; s ^= s >> 7; s ^= s << 17; return s; }

static uint256 randU(int maxbits)
{
    uint256 r;
    unsigned char* p = r.begin();
    for (int i = 0; i < 32; i++) p[i] = (unsigned char)xr();
    r >>= (256 - (int)(xr() % (maxbits + 1)));
    return r;
}

static std::string hex512(const uint512& w)
{
    const unsigned char* p = (const unsigned char*)w.begin();
    char buf[129];
    for (int i = 0; i < 64; i++) sprintf(buf + i * 2, "%02x", p[63 - i]);
    buf[128] = 0;
    return buf;
}

int main()
{
    // Deterministic edge cases first.
    {
        uint32_t edgeBits[] = {
            0x1d00ffff, 0x1e0fffff, 0x1f00ffff, 0x1f7fffff, 0x20000001,
            0x03000000, 0x01003456, 0x02008000, 0x05009234, 0x1b0404cb,
            0xff123456, 0x21010000, 0x22000100, 0x1d008000,
        };
        for (unsigned e = 0; e < sizeof(edgeBits)/sizeof(edgeBits[0]); e++) {
            uint256 c; bool neg=false, over=false;
            c.SetCompact(edgeBits[e], &neg, &over);
            printf("SETCOMPACT %08x %s %d %d\n", edgeBits[e], c.GetHex().c_str(), (int)neg, (int)over);
            if (!over && !(c == uint256(0)))
                printf("GETCOMPACT %s %08x\n", c.GetHex().c_str(), c.GetCompact(neg));
        }
        uint256 one(1), zero(0), big; big = ~uint256(0);
        printf("MUL %s %s %s\n", big.GetHex().c_str(), one.GetHex().c_str(), (big*one).GetHex().c_str());
        printf("MUL %s %s %s\n", big.GetHex().c_str(), zero.GetHex().c_str(), (big*zero).GetHex().c_str());
        printf("MUL %s %s %s\n", big.GetHex().c_str(), big.GetHex().c_str(), (big*big).GetHex().c_str());
        printf("MULW %s %s %s\n", big.GetHex().c_str(), big.GetHex().c_str(), hex512(uint512(big)*uint512(big)).c_str());
        printf("DIV %s %s %s\n", big.GetHex().c_str(), one.GetHex().c_str(), (big/one).GetHex().c_str());
        printf("DIV %s %s %s\n", one.GetHex().c_str(), big.GetHex().c_str(), (one/big).GetHex().c_str());
        uint256 seven(7), three(3);
        printf("DIV %s %s %s\n", seven.GetHex().c_str(), three.GetHex().c_str(), (seven/three).GetHex().c_str());
    }

    for (int i = 0; i < 60000; i++) {
        uint256 a = randU(256), b = randU(256);
        printf("MUL %s %s %s\n", a.GetHex().c_str(), b.GetHex().c_str(), (a * b).GetHex().c_str());
        printf("MULW %s %s %s\n", a.GetHex().c_str(), b.GetHex().c_str(),
               hex512(uint512(a) * uint512(b)).c_str());
        if (!(b == uint256(0)))
            printf("DIV %s %s %s\n", a.GetHex().c_str(), b.GetHex().c_str(), (a / b).GetHex().c_str());

        // compact round-trips
        uint32_t nBits = ((1 + (uint32_t)(xr() % 33)) << 24) | (uint32_t)(xr() & 0x7fffff);
        uint256 c; bool neg = false, over = false;
        c.SetCompact(nBits, &neg, &over);
        printf("SETCOMPACT %08x %s %d %d\n", nBits, c.GetHex().c_str(), (int)neg, (int)over);
        if (!over && !(c == uint256(0)))
            printf("GETCOMPACT %s %08x\n", c.GetHex().c_str(), c.GetCompact(neg));

        // wide retarget: (a * m) / d  narrowed to 256
        uint256 tgt; bool o2 = false;
        tgt.SetCompact(((1 + (uint32_t)(xr() % 31)) << 24) | (uint32_t)(xr() & 0x7fffff), 0, &o2);
        if (!o2) {
            int64_t m = 1 + (int64_t)(xr() % 8000000);
            int64_t d = 1 + (int64_t)(xr() % 8000000);
            uint512 w = uint512(tgt) * uint512((uint64_t)m);
            w /= uint512((uint64_t)d);
            printf("RETARGET %s %lld %lld %s\n", tgt.GetHex().c_str(),
                   (long long)m, (long long)d, w.trim256().GetHex().c_str());

            // block trust: PoS 2^256/(t+1), PoW powLimit/(t+1)
            if (!(tgt == uint256(0))) {
                uint256 one(1);
                uint256 pos = (uint256(~tgt) / (tgt + one)) + one;
                uint256 powLimit = ~uint256(0); powLimit >>= 8;
                uint256 pw = powLimit / (tgt + one);
                printf("TRUST %s %s %s\n", tgt.GetHex().c_str(), pos.GetHex().c_str(), pw.GetHex().c_str());
            }
        }
    }
    return 0;
}
